sw/device/tests/spi_device_flash_smoketest.c

```diff
@@ -0,0 +1,166 @@
+      },
+      {
+          // Slot 1: ReadStatus2
+          .opcode = kSpiDeviceFlashOpReadStatus2,
+          .address_type = kDifSpiDeviceFlashAddrDisabled,
+          .dummy_cycles = 0,
+          .payload_io_type = kDifSpiDevicePayloadIoSingle,
+          .payload_dir_to_host = true,
+      },
+      {
+          // Slot 2: ReadStatus3
+          .opcode = kSpiDeviceFlashOpReadStatus3,
+          .address_type = kDifSpiDeviceFlashAddrDisabled,
+          .dummy_cycles = 0,
+          .payload_io_type = kDifSpiDevicePayloadIoSingle,
+          .payload_dir_to_host = true,
+      },
+      {
+          // Slot 3: ReadJedecID
+          .opcode = kSpiDeviceFlashOpReadJedec,
+          .address_type = kDifSpiDeviceFlashAddrDisabled,
+          .dummy_cycles = 0,
+          .payload_io_type = kDifSpiDevicePayloadIoSingle,
+          .payload_dir_to_host = true,
+      },
+      {
+          // Slot 4: ReadSfdp
+          .opcode = kSpiDeviceFlashOpReadSfdp,
+          .address_type = kDifSpiDeviceFlashAddr3Byte,
+          .dummy_cycles = 8,
+          .payload_io_type = kDifSpiDevicePayloadIoSingle,
+          .payload_dir_to_host = true,
+      },
+      {
+          // Slot 1: ReadNormal
+          .opcode = kSpiDeviceFlashOpReadNormal,
+          .address_type = kDifSpiDeviceFlashAddr3Byte,
+          .dummy_cycles = 0,
+          .payload_io_type = kDifSpiDevicePayloadIoSingle,
+          .payload_dir_to_host = true,
+      },
+  };
+
+  for (uint8_t i = 0; i < ARRAYSIZE(read_commands); ++i) {
+    uint8_t slot = i + kSpiDeviceReadCommandSlotBase;
+    TRY(dif_spi_device_set_flash_command_slot(spid, slot, kDifToggleEnabled,
+                                              read_commands[i]));
+  }
+
+  dif_spi_device_flash_command_t write_commands[] = {
+      {
+          // Slot 1: PageProgram
+          .opcode = kSpiDeviceFlashOpPageProgram,
+          .address_type = kDifSpiDeviceFlashAddrCfg,
+          .payload_io_type = kDifSpiDevicePayloadIoSingle,
+          .payload_dir_to_host = false,
+          .upload = true,
+          .set_busy_status = true,
+      },
+  };
+
+  for (uint8_t i = 0; i < ARRAYSIZE(write_commands); ++i) {
+    uint8_t slot = i + (uint8_t)kSpiDeviceWriteCommandSlotBase;
+    TRY(dif_spi_device_set_flash_command_slot(spid, slot, kDifToggleEnabled,
+                                              write_commands[i]));
+  }
+  return OK_STATUS();
+}
+
+bool test_main(void) {
+  dif_spi_device_handle_t spid;
+  CHECK_DIF_OK(dif_spi_device_init_handle(
+      mmio_region_from_addr(TOP_EARLGREY_SPI_DEVICE_BASE_ADDR), &spid));
+
+  CHECK_STATUS_OK(configure_flash_mode(&spid));
+
+  LOG_INFO("SYNC: Waiting for write page");
+  upload_info_t info;
+  CHECK_STATUS_OK(spi_device_testutils_wait_for_upload(&spid, &info));
+
+  CHECK_ARRAYS_EQ(info.data, kSpiTxData, ARRAYSIZE(kSpiTxData));
+  CHECK_DIF_OK(dif_spi_device_set_flash_status_registers(&spid, 0x00));
+
+  CHECK_DIF_OK(dif_spi_device_write_flash_buffer(
```

Comment:
You might consider swapping these two. Then there's no need for the UART SYNC, since the BUSY bit dropping would be the synchronization.

It's okay to leave in the UART message, though, if you feel it would aid debugging.

Reply:
Good catch. Done.